File: products/yappc/core/framework/src/main/java/com/ghatana/yappc/framework/core/config/FeatureFlags.java
package com.ghatana.yappc.framework.core.config;

import io.activej.config.Config;
import io.activej.config.converter.ConfigConverters;
import io.activej.inject.annotation.Inject;
import io.activej.inject.annotation.Provides;
import org.jetbrains.annotations.NotNull;

/**
 * Feature flags configuration for YAPPC platform.
 *
 * <p>Provides feature flag management for gradual rollout of incomplete features.
 * Features can be enabled/disabled via configuration without code changes.</p>
 *
 * @doc.type class
 * @doc.purpose Feature flag management for gradual rollout
 * @doc.layer infrastructure
 * @doc.pattern Configuration
 */
public final class FeatureFlags {

    // Singleton instance for static access (set by DI or explicitly)
    private static volatile FeatureFlags INSTANCE;

    /**
     * Set the global FeatureFlags instance.
     * Called during application initialization.
     */
    public static void setGlobal(@NotNull FeatureFlags flags) {
        INSTANCE = flags;
    }

    /**
     * Static convenience method to check if a feature flag is enabled.
     * Returns false if no global instance has been set.
     *
     * @param flag the feature flag to check
     * @return true if enabled, false otherwise
     */
    public static boolean isEnabled(@NotNull FeatureFlag flag) {
        FeatureFlags instance = INSTANCE;
        if (instance == null) {
            return false; // Safe default when not yet initialized
        }
        return instance.isFeatureEnabled(flag.key());
    }

    // LLM Provider Features
    public static final String LLM_OPENAI_ENABLED = "feature.llm.openai.enabled";
    public static final String LLM_ANTHROPIC_ENABLED = "feature.llm.anthropic.enabled";
    public static final String LLM_OLLAMA_ENABLED = "feature.llm.ollama.enabled";

    // AI Features
    public static final String AI_CANVAS_GENERATION = "feature.ai.canvas.generation";
    public static final String AI_CODE_REVIEW = "feature.ai.code.review";
    public static final String AI_REQUIREMENTS_PARSING = "feature.ai.requirements.parsing";

    // DataCloud Features
    public static final String DATACLOUD_DASHBOARD = "feature.datacloud.dashboard";
    public static final String DATACLOUD_WIDGET = "feature.datacloud.widget";

    // UI Features
    public static final String UI_CANVAS_EDITING = "feature.ui.canvas.editing";
    public static final String UI_DARK_MODE = "feature.ui.dark.mode";
    public static final String UI_MOBILE_VIEW = "feature.ui.mobile.view";

    // Integration Features
    public static final String INTEGRATION_GITHUB = "feature.integration.github";
    public static final String INTEGRATION_FIGMA = "feature.integration.figma";

    // Development Features
    public static final String DEV_DEBUG_MODE = "feature.dev.debug";
    public static final String DEV_MOCK_SERVICES = "feature.dev.mock.services";

    private final Config config;

    @Inject
    private FeatureFlags(@NotNull Config config) {
        this.config = config;
    }

    @Provides
    public static FeatureFlags create(@NotNull Config config) {
        return new FeatureFlags(config);
    }

    // LLM Provider Checks

    /**
     * Check if OpenAI provider is enabled.
     */
    public boolean isOpenAIEnabled() {
        return config.get(ConfigConverters.ofBoolean(), LLM_OPENAI_ENABLED, false);
    }

    /**
     * Check if Anthropic provider is enabled.
     */
    public boolean isAnthropicEnabled() {
        return config.get(ConfigConverters.ofBoolean(), LLM_ANTHROPIC_ENABLED, false);
    }

    /**
     * Check if Ollama provider is enabled.
     */
    public boolean isOllamaEnabled() {
        return config.get(ConfigConverters.ofBoolean(), LLM_OLLAMA_ENABLED, false);
    }

    /**
     * Get the default LLM provider to use.
     */
    @NotNull
    public String getDefaultLLMProvider() {
        if (isOpenAIEnabled()) {
            return "openai";
        } else if (isAnthropicEnabled()) {
            return "anthropic";
        } else if (isOllamaEnabled()) {
            return "ollama";
        }
        return "mock";
    }

    // AI Feature Checks

    /**
     * Check if canvas AI generation is enabled.
     */
    public boolean isCanvasGenerationEnabled() {
        return config.get(ConfigConverters.ofBoolean(), AI_CANVAS_GENERATION, false);
    }

    /**
     * Check if AI code review is enabled.
     */
    public boolean isCodeReviewEnabled() {
        return config.get(ConfigConverters.ofBoolean(), AI_CODE_REVIEW, false);
    }

    /**
     * Check if AI requirements parsing is enabled.
     */
    public boolean isRequirementsParsingEnabled() {
        return config.get(ConfigConverters.ofBoolean(), AI_REQUIREMENTS_PARSING, false);
    }

    // DataCloud Feature Checks

    /**
     * Check if DataCloud dashboard persistence is enabled.
     */
    public boolean isDashboardPersistenceEnabled() {
        return config.get(ConfigConverters.ofBoolean(), DATACLOUD_DASHBOARD, false);
    }

    /**
     * Check if DataCloud widget persistence is enabled.
     */
    public boolean isWidgetPersistenceEnabled() {
        return config.get(ConfigConverters.ofBoolean(), DATACLOUD_WIDGET, false);
    }

    // UI Feature Checks

    /**
     * Check if canvas editing UI is enabled.
     */
    public boolean isCanvasEditingEnabled() {
        return config.get(ConfigConverters.ofBoolean(), UI_CANVAS_EDITING, true);
    }

    /**
     * Check if dark mode is enabled.
     */
    public boolean isDarkModeEnabled() {
        return config.get(ConfigConverters.ofBoolean(), UI_DARK_MODE, true);
    }

    /**
     * Check if mobile view is enabled.
     */
    public boolean isMobileViewEnabled() {
        return config.get(ConfigConverters.ofBoolean(), UI_MOBILE_VIEW, false);
    }

    // Integration Feature Checks

    /**
     * Check if GitHub integration is enabled.
     */
    public boolean isGitHubIntegrationEnabled() {
        return config.get(ConfigConverters.ofBoolean(), INTEGRATION_GITHUB, false);
    }

    /**
     * Check if Figma integration is enabled.
     */
    public boolean isFigmaIntegrationEnabled() {
        return config.get(ConfigConverters.ofBoolean(), INTEGRATION_FIGMA, false);
    }

    // Development Feature Checks

    /**
     * Check if debug mode is enabled.
     */
    public boolean isDebugModeEnabled() {
        return config.get(ConfigConverters.ofBoolean(), DEV_DEBUG_MODE, false);
    }

    /**
     * Check if mock services are enabled (for development/testing).
     */
    public boolean areMockServicesEnabled() {
        return config.get(ConfigConverters.ofBoolean(), DEV_MOCK_SERVICES, true);
    }

    /**
     * Check if a feature is enabled by name.
     */
    public boolean isFeatureEnabled(@NotNull String featureName) {
        return config.get(ConfigConverters.ofBoolean(), featureName, false);
    }

    /**
     * Get a feature flag value with default.
     */
    @NotNull
    public String getFeatureValue(@NotNull String featureName, @NotNull String defaultValue) {
        return config.get(featureName, defaultValue);
    }
}

File: products/yappc/frontend/libs/config/src/features/feature-flags.ts
/**
 * Feature Flags Configuration for YAPPC App Creator
 *
 * Provides feature flag management for gradual rollout of incomplete features.
 * Features can be enabled/disabled via environment variables or runtime config.
 *
 * @doc.type module
 * @doc.purpose Feature flag management for frontend
 * @doc.layer infrastructure
 * @doc.pattern Configuration
 */

// Feature Flag Names
export const FEATURE_FLAGS = {
    // LLM Provider Features
    LLM_OPENAI_ENABLED: 'feature.llm.openai.enabled',
    LLM_ANTHROPIC_ENABLED: 'feature.llm.anthropic.enabled',
    LLM_OLLAMA_ENABLED: 'feature.llm.ollama.enabled',

    // AI Features
    AI_CANVAS_GENERATION: 'feature.ai.canvas.generation',
    AI_CODE_REVIEW: 'feature.ai.code.review',
    AI_REQUIREMENTS_PARSING: 'feature.ai.requirements.parsing',

    // UI Features
    UI_CANVAS_EDITING: 'feature.ui.canvas.editing',
    UI_DARK_MODE: 'feature.ui.dark.mode',
    UI_MOBILE_VIEW: 'feature.ui.mobile.view',
    UI_COMMAND_PALETTE: 'feature.ui.command.palette',
    UI_KEYBOARD_SHORTCUTS: 'feature.ui.keyboard.shortcuts',

    // Integration Features
    INTEGRATION_GITHUB: 'feature.integration.github',
    INTEGRATION_FIGMA: 'feature.integration.figma',

    // Development Features
    DEV_DEBUG_MODE: 'feature.dev.debug',
    DEV_MOCK_SERVICES: 'feature.dev.mock.services',
    DEV_PERFORMANCE_MONITOR: 'feature.dev.performance.monitor',
} as const;

export type FeatureFlagName = typeof FEATURE_FLAGS[keyof typeof FEATURE_FLAGS];

/**
 * Default feature flag values
 */
const DEFAULT_FLAGS: Record<FeatureFlagName, boolean> = {
    // LLM Providers - Off by default, enable via env
    [FEATURE_FLAGS.LLM_OPENAI_ENABLED]: false,
    [FEATURE_FLAGS.LLM_ANTHROPIC_ENABLED]: false,
    [FEATURE_FLAGS.LLM_OLLAMA_ENABLED]: false,

    // AI Features - Off by default
    [FEATURE_FLAGS.AI_CANVAS_GENERATION]: false,
    [FEATURE_FLAGS.AI_CODE_REVIEW]: false,
    [FEATURE_FLAGS.AI_REQUIREMENTS_PARSING]: false,

    // UI Features - On by default
    [FEATURE_FLAGS.UI_CANVAS_EDITING]: true,
    [FEATURE_FLAGS.UI_DARK_MODE]: true,
    [FEATURE_FLAGS.UI_MOBILE_VIEW]: false,
    [FEATURE_FLAGS.UI_COMMAND_PALETTE]: true,
    [FEATURE_FLAGS.UI_KEYBOARD_SHORTCUTS]: true,

    // Integration Features - Off by default
    [FEATURE_FLAGS.INTEGRATION_GITHUB]: false,
    [FEATURE_FLAGS.INTEGRATION_FIGMA]: false,

    // Development Features - Environment-dependent
    [FEATURE_FLAGS.DEV_DEBUG_MODE]: import.meta.env.DEV,
    [FEATURE_FLAGS.DEV_MOCK_SERVICES]: import.meta.env.DEV,
    [FEATURE_FLAGS.DEV_PERFORMANCE_MONITOR]: false,
};

/**
 * Feature flag configuration interface
 */
export interface FeatureFlagConfig {
    flags: Record<FeatureFlagName, boolean>;
    overrides: Record<string, boolean>;
}

/**
 * Get feature flag value from environment variable
 */
function getEnvFlag(name: string): boolean | undefined {
    const envValue = import.meta.env[`VITE_${name.replace(/\./g, '_').toUpperCase()}`];
    if (envValue === 'true' || envValue === '1') return true;
    if (envValue === 'false' || envValue === '0') return false;
    return undefined;
}

/**
 * Feature Flags Hook Return Type
 */
export interface UseFeatureFlagsReturn {
    isEnabled: (flag: FeatureFlagName) => boolean;
    getValue: (flag: FeatureFlagName) => boolean;
    setOverride: (flag: string, value: boolean) => void;
    clearOverride: (flag: string) => void;
    isOverridden: (flag: string) => boolean;
    getAllFlags: () => Record<FeatureFlagName, boolean>;
    reload: () => void;
}

/**
 * Create feature flags context value
 */
export function createFeatureFlags(): UseFeatureFlagsReturn {
    // Runtime overrides (can be set by user preferences or testing)
    const overrides = new Map<string, boolean>();

    // Initialize flags with environment variable overrides
    const getInitialValue = (flag: FeatureFlagName): boolean => {
        // Check runtime override first
        if (overrides.has(flag)) {
            return overrides.get(flag)!;
        }

        // Check environment variable
        const envValue = getEnvFlag(flag);
        if (envValue !== undefined) {
            return envValue;
        }

        // Return default
        return DEFAULT_FLAGS[flag];
    };

    const getAllFlagsInternal = (): Record<FeatureFlagName, boolean> => {
        const flags = {} as Record<FeatureFlagName, boolean>;
        for (const flag of Object.values(FEATURE_FLAGS) as FeatureFlagName[]) {
            flags[flag] = getInitialValue(flag);
        }
        return flags;
    };

    return {
        isEnabled: (flag: FeatureFlagName): boolean => {
            return getInitialValue(flag);
        },

        getValue: (flag: FeatureFlagName): boolean => {
            return getInitialValue(flag);
        },

        setOverride: (flag: string, value: boolean): void => {
            overrides.set(flag, value);
        },

        clearOverride: (flag: string): void => {
            overrides.delete(flag);
        },

        isOverridden: (flag: string): boolean => {
            return overrides.has(flag);
        },

        getAllFlags: (): Record<FeatureFlagName, boolean> => {
            return getAllFlagsInternal();
        },

        reload: (): void => {
            // Clear cache if needed - triggers re-read of env vars
            // For now, this just triggers a re-computation
        },
    };
}

/**
 * React Context for Feature Flags
 */
export const FeatureFlagsContext = createContext<UseFeatureFlagsReturn | null>(null);

/**
 * Provider component for feature flags
 */
export function FeatureFlagsProvider({
    children,
}: {
    children: React.ReactNode;
}): JSX.Element {
    const featureFlags = useMemo(() => createFeatureFlags(), []);

    return (
        <FeatureFlagsContext.Provider value= { featureFlags } >
        { children }
        </FeatureFlagsContext.Provider>
  );
}

/**
 * Hook to access feature flags
 */
export function useFeatureFlags(): UseFeatureFlagsReturn {
    const context = useContext(FeatureFlagsContext);
    if (!context) {
        throw new Error('useFeatureFlags must be used within FeatureFlagsProvider');
    }
    return context;
}

/**
 * Hook to check if a specific feature is enabled
 */
export function useFeatureEnabled(flag: FeatureFlagName): boolean {
    const { isEnabled } = useFeatureFlags();
    return isEnabled(flag);
}

/**
 * Conditional component that only renders if feature is enabled
 */
export function Feature({
    flag,
    children,
    fallback = null,
}: {
    flag: FeatureFlagName;
    children: React.ReactNode;
    fallback?: React.ReactNode;
}): JSX.Element | null {
    const isEnabled = useFeatureEnabled(flag);
    return isEnabled ? <>{ children } < /> : <>{fallback}</ >;
}

/**
 * Feature flag configuration for external config files
 */
export const FEATURE_CONFIG_TEMPLATE = {
    features: {
        llm: {
            openai: false,
            anthropic: false,
            ollama: false,
        },
        ai: {
            canvasGeneration: false,
            codeReview: false,
            requirementsParsing: false,
        },
        ui: {
            canvasEditing: true,
            darkMode: true,
            mobileView: false,
            commandPalette: true,
            keyboardShortcuts: true,
        },
        integrations: {
            github: false,
            figma: false,
        },
        dev: {
            debugMode: import.meta.env.DEV,
            mockServices: import.meta.env.DEV,
            performanceMonitor: false,
        },
    },
};
